### PR TITLE
add labels, ejectables & other fixes

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/StackManager.js [hunk 1]
```diff
@@ -72,21 +76,26 @@ export default class StackManager {
     const file = dockerComposePath(this.getProjectPath());
     const command = clean(args).join(' ');
 
-    return exec(`docker-compose --file ${file} ${command}`);
+    return exec(`docker-compose -f ${file} -p ${this.getProjectName()} ${command}`);
```
Comment (deiucanta) [inline src/StackManager.js]: we should make the signature of `exec` similar to `spawn`: command, args and options

File: src/StackManager.js [hunk 2]
```diff
@@ -72,21 +76,26 @@ export default class StackManager {
...
     // cancel default SIGINT behaviour
     process.on('SIGINT', () => {});
 
     return child;
   }
-  async listContainers(serviceName = null) {
-    const stdout = await this.execDockerCompose('ps -q', serviceName);
+  async listRunningContainers(serviceName) {
+    const label = `${this.getProjectName()}.${serviceName}`;
+    const stdout = await this.execDocker('ps -q', '--filter', `label=stacker=${label}`, '--filter', 'status=running');
```
Comment (deiucanta) [inline src/StackManager.js]: I would move the `clean` action before the call, outside of `execDocker` or `execDockerCompose`. This way, the author of a method will be aware on how things are supposed to work.

We could have the following method signatures

- utils: `exec(command, args, options)`
- utils: `spawn(command, args, options)`
- here: `execDocker(subcommand, args)`
- here: `execDockerCompose(subcommand, args)`
- here: `spawnDockerCompose(subcommand, args)`

Then, inside the `listContainers` we will call `execDocker('ps', clean([...]))`

Comment (deiucanta) [reply src/StackManager.js]: Also, when the arguments ar inside an array we can arrange them nicely on multiple lines (if needed).

File: src/templates/laravel/LaravelStack.js [hunk 3]
```diff
@@ -48,7 +50,13 @@ export default class LaravelStack extends Stack {
...
-      path: '/etc/apache2/site-available/000-default.conf',
+      path: '/etc/apache2/site-available/app.conf',
+      service: 'laravel',
+    });
+
+    this.runnables.set('apache-reload', {
+      label: 'Reload Apache service',
+      exec: 'service apache2 reload',
       service: 'laravel',
     });
```
Comment (deiucanta) [inline src/templates/laravel/LaravelStack.js:61]: Another runnable we could have on `laravel` is `phpunit`. A command that will run the tests.

Not sure if `phpunit` is available inside our container image. If not, let's queue this up in Trello and handle it later.

File: src/StackManager.js (hunk 2 above)
Comment (ghost) [reply src/StackManager.js]: The `execDockerCompose` method is not used

Comment (deiucanta) [reply src/StackManager.js]: then, let's remove it

File: src/StackManager.js [hunk 4]
```diff
@@ -138,37 +148,50 @@ export default class StackManager {
...
     const shell = this.stack.services.get(serviceName).shell;
 
     if (!shell) throw new Error('This service does not have any shell.');
 
-    return this.spawnDockerCompose('exec', serviceName, shell, '-c', runnable.exec);
+    const containers = await this.listRunningContainers(serviceName);
+
+    if (!containers.length) throw new Error('You need to start the project first. Run "stacker up".');
+
+    return this.spawnDockerCompose('exec', clean([serviceName, shell]));
```
Comment (deiucanta) [inline src/StackManager.js]: no need for `clean` here

File: src/StackManager.js [hunk 5]
```diff
@@ -182,16 +205,9 @@ export default class StackManager {
...
-  shell(serviceName) {
-    const shell = this.stack.services.get(serviceName).shell;
-
-    if (!shell) throw new Error('This service does not have any shell.');
-
-    return this.spawnDockerCompose('exec', serviceName, shell);
+    return this.execDocker('cp', clean([
+      '--follow-link',
+      `${container}:${remotePath} ${localPath}`,
+    ]));
```
Comment (deiucanta) [inline src/StackManager.js]: - same here, we should use `clean` only when we have some optional arguments
- also, we can split the 2nd item and have `localPath` as the 3rd param

File: src/StackManager.js [hunk 6]
```diff
@@ -63,30 +67,35 @@ export default class StackManager {
...
     process.on('SIGINT', () => {});
 
     return child;
   }
-  async listContainers(serviceName = null) {
-    const stdout = await this.execDockerCompose('ps -q', serviceName);
+  async listRunningContainers(serviceName) {
+    const label = `${this.getProjectName()}.${serviceName}`;
+    const stdout = await this.execDocker('ps', clean([
+      '-q',
```
Comment (deiucanta) [inline src/StackManager.js]: since line width is not a concern, let's have full option names for better readability

**Pull request merged**